nvflare/app_common/ccwf/server_ctl.py

```diff
@@ -0,0 +1,459 @@
+            AppConstants.NUM_ROUNDS: self.num_rounds,
+            Constant.START_ROUND: self.start_round,
+            FLContextKey.WORKFLOW: self.workflow_id,
+        }
+
+        extra_config = self.prepare_config()
+        if extra_config:
+            learn_config.update(extra_config)
+
+        # configure all clients
+        shareable = Shareable()
+        shareable[Constant.CONFIG] = learn_config
+
+        task = Task(
+            name=self.configure_task_name,
+            data=shareable,
+            timeout=self.configure_task_timeout,
+            result_received_cb=self._process_configure_reply,
+        )
+
+        self.log_info(fl_ctx, f"sending task {self.configure_task_name} to clients {self.participating_clients}")
+        start_time = time.time()
+        self.broadcast_and_wait(
+            task=task,
+            targets=self.participating_clients,
+            min_responses=len(self.participating_clients),
+            fl_ctx=fl_ctx,
+            abort_signal=abort_signal,
+        )
+
+        time_taken = time.time() - start_time
+        self.log_info(fl_ctx, f"client configuration took {time_taken} seconds")
+
+        failed_clients = []
+        for c, cs in self.client_statuses.items():
+            assert isinstance(cs, ClientStatus)
+            if not cs.ready_time:
+                failed_clients.append(c)
+
+        if failed_clients:
+            self.system_panic(
+                f"failed to configure clients {failed_clients}",
+                fl_ctx,
+            )
+            return
+
+        self.log_info(fl_ctx, f"successfully configured clients {self.participating_clients}")
+
+        # starting the starting_client
+        if self.starting_client:
+            shareable = Shareable()
+            task = Task(
+                name=self.start_task_name,
+                data=shareable,
+                timeout=self.start_task_timeout,
+                result_received_cb=self._process_start_reply,
+            )
+
+            self.log_info(fl_ctx, f"sending task {self.start_task_name} to client {self.starting_client}")
+
+            self.send_and_wait(
+                task=task,
+                targets=[self.starting_client],
+                fl_ctx=fl_ctx,
+                abort_signal=abort_signal,
+            )
+
+            if not self.cw_started:
+                self.system_panic(
+                    f"failed to start workflow {self.workflow_id} on client {self.starting_client}",
+                    fl_ctx,
+                )
+                return
+
+            self.log_info(fl_ctx, f"started workflow {self.workflow_id} on client {self.starting_client}")
+
+        # a subclass could provide additional control flow
+        self.sub_flow(abort_signal, fl_ctx)
+
+        self.log_info(fl_ctx, f"Waiting for clients to finish workflow {self.workflow_id}  ...")
+        while not abort_signal.triggered and not self.asked_to_stop:
```

Comment:
if self.starting_client is not set and sub_flow is not implemented, we will stuck at L275.

Reply:
That would be the problem with the subclass that implements the concrete workflow.
In this case, the system is not stuck - it is idle and useless. You can still end it by aborting the job.